### PR TITLE
BI-2127 Position Information missing from exp download file

### PR DESCRIPTION
**Pull request by davedrp**
# Description
[BI-2127](https://breedinginsight.atlassian.net/browse/BI-2127) Position Information missing from exp download file

There were 2 separate problems:

1. The `row` and `column` fields would only populate if there was data for both `row` **and** `column`.  If a record only had data for `row` but not `column` or If a record only had data for `column` but not `row`, then the data for neither `row` or `column`would appear for that record.
2. The data for `Lat`, 'Long`, `Elevation`, and `RTK` was not being added to the download spreadsheet.

# Testing

1. Import an experiment with `row` data but not `column` data.
2. Import an experiment with `column` data but not `row` data.
3. In the above two import files, include data for `Lat`, 'Long`, `Elevation`, and `RTK` 
4. Download the experiments

**EXPECTED RESULTS**
The download data should match the import data.

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have tested my code and ensured it meets the acceptance criteria of the story
- [ ] I have tested that my code works with both the brapi-java-server and BreedBase
- [ ] I have create/modified unit tests to cover this change
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to documentation
- [ ] I have run TAF: _\<please include a link to TAF run>_

[BI-2127]: https://breedinginsight.atlassian.net/browse/BI-2127?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by davedrp** [comment]
Sample file for testing
[short_noCol_exp.xlsx](https://github.com/Breeding-Insight/bi-api/files/15309874/short_noCol_exp.xlsx)
[short_noRow_exp.xlsx](https://github.com/Breeding-Insight/bi-api/files/15309875/short_noRow_exp.xlsx)
[short_noRowOrCol_exp.xlsx](https://github.com/Breeding-Insight/bi-api/files/15309876/short_noRowOrCol_exp.xlsx)
[short_RowAndCol_exp.xlsx](https://github.com/Breeding-Insight/bi-api/files/15309877/short_RowAndCol_exp.xlsx)

File: src/main/java/org/breedinginsight/brapi/v2/services/BrAPITrialService.java [hunk 1]
```diff
@@ -484,11 +501,13 @@ private Map<String, Object> createExportRow(
                 .findFirst();
         blockLevel.ifPresent(brAPIObservationUnitLevelRelationship ->
                 row.put(ExperimentObservation.Columns.BLOCK_NUM, Integer.parseInt(brAPIObservationUnitLevelRelationship.getLevelCode())));
-        if (ou.getObservationUnitPosition() != null && ou.getObservationUnitPosition().getPositionCoordinateX() != null &&
-                ou.getObservationUnitPosition().getPositionCoordinateY() != null) {
+
+        //Row and Column
+        if ( ou.getObservationUnitPosition() != null ) {
```
Comment (mlm483) [inline src/main/java/org/breedinginsight/brapi/v2/services/BrAPITrialService.java:506]: Is it allowed to upload a file that has row but not column data or vice versa? I think this change makes sense regardless.

File: src/main/java/org/breedinginsight/brapi/v2/services/BrAPITrialService.java [hunk 2]
```diff
@@ -499,7 +518,25 @@ private Map<String, Object> createExportRow(
         return row;
     }
 
-
+    private String doubleToString(double val){
+        return Double.isNaN(val) ? null : String.valueOf( val );
```
Comment (mlm483) [inline src/main/java/org/breedinginsight/brapi/v2/services/BrAPITrialService.java:522]: Nice.

**Review by mlm483 — APPROVED** [review]

File: src/main/java/org/breedinginsight/brapi/v2/services/BrAPITrialService.java [hunk 3]
```diff
@@ -465,6 +468,20 @@ private Map<String, Object> createExportRow(
         row.put(ExperimentObservation.Columns.EXP_TYPE, experiment.getAdditionalInfo().getAsJsonObject().get(BrAPIAdditionalInfoFields.EXPERIMENT_TYPE).getAsString());
         row.put(ExperimentObservation.Columns.ENV, Utilities.removeProgramKeyAndUnknownAdditionalData(study.getStudyName(), program.getKey()));
         row.put(ExperimentObservation.Columns.ENV_LOCATION, Utilities.removeProgramKey(study.getLocationName(), program.getKey()));
+
+        // Lat, Long, Elevation
+        Coordinates coordinates = extractCoordinates(ou);
+        row.put( ExperimentObservation.Columns.LAT, coordinates==null? null : doubleToString(coordinates.getLat()) );
+        row.put( ExperimentObservation.Columns.LONG, coordinates==null? null : doubleToString(coordinates.getLon()) );
+        row.put( ExperimentObservation.Columns.ELEVATION, coordinates==null? null : doubleToString(coordinates.getAlt()) );
```
Comment (dmeidlin) [inline src/main/java/org/breedinginsight/brapi/v2/services/BrAPITrialService.java:476]: Looks good! Refactoring using Optionals here might help the readability.
```suggestion
    // Lat, Long, Elevation
    Coordinates coordinates = extractCoordinates(ou);
    Optional.ofNullable(coordinates)
            .map(c -> doubleToString(c.getLat()))
            .ifPresent(lat -> row.put(ExperimentObservation.Columns.LAT, lat));
    Optional.ofNullable(coordinates)
            .map(c -> doubleToString(c.getLon()))
            .ifPresent(lon -> row.put(ExperimentObservation.Columns.LONG, lon));
    Optional.ofNullable(coordinates)
            .map(c -> doubleToString(c.getAlt()))
            .ifPresent(elevation -> row.put(ExperimentObservation.Columns.ELEVATION, elevation));
```

**Review by dmeidlin — APPROVED** [review]

**Pull request merged**